src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs

```diff
@@ -118,46 +127,34 @@ internal DocumentInfo(DocumentAttributes attributes, TextLoader? loader, IDocume
         }
 
         public DocumentInfo WithId(DocumentId id)
-        {
-            return With(attributes: Attributes.With(id: id));
-        }
+            => With(attributes: Attributes.With(id: id ?? throw new ArgumentNullException(nameof(id))));
 
         public DocumentInfo WithName(string name)
-        {
-            return this.With(attributes: Attributes.With(name: name));
-        }
+            => With(attributes: Attributes.With(name: name ?? throw new ArgumentNullException(nameof(name))));
 
         public DocumentInfo WithFolders(IEnumerable<string>? folders)
-        {
-            return this.With(attributes: Attributes.With(folders: folders.ToImmutableReadOnlyListOrEmpty()));
-        }
+            => With(attributes: Attributes.With(folders: folders.AsBoxedImmutableArrayWithNonNullItems() ?? throw new ArgumentNullException(nameof(folders))));
```

Comment:
is this correct?  looks like we doc'ed that it could be null.

Reply:
It's quite inconsistent. 

`Solution.WirhtDocumentFolders` filters out `null` values. This API does not. 

Hmm, it seems that we handle null values ok, although our nullable annotations say otherwise. So perhaps we should just keep nulls in the array (not filter them out in `Solution.WirhtDocumentFolders`) and fix the annotations up.

Reply:
I'm gonna remove the filtering and throw if there are non-null items.